staging/src/k8s.io/apimachinery/pkg/watch/watch.go

```diff
@@ -140,40 +146,40 @@ func (a *AggregatedWatcher) AddWatchInterface(watcher Interface, err error) {
 
 			for {
 				select {
-				case <-stopCh.Read:
-					a.closeWatcher(w, stopCh)
+				case <-stopch:
+					a.closeWatcher(w)
 					return
 				case signal, ok := <-w.ResultChan():
 					if !ok {
 						//klog.Infof("watch channel %v closed for aggregated chan %#v.", w, a.aggChan)
-						a.closeWatcher(w, stopCh)
+						a.closeWatcher(w)
 						return
-					} else {
-						//klog.V(3).Infof("Get event (chan %#v) %s.", a.aggChan, PrintEvent(signal))
 					}
 
 					select {
-					case <-stopCh.Read:
-						a.closeWatcher(w, stopCh)
+					case <-stopch:
+						a.closeWatcher(w)
 						return
 					case a.aggChan <- signal:
 						//klog.V(3).Infof("Sent event (chan %#v) %s.", a.aggChan, PrintEvent(signal))
 					}
 				}
 			}
-		}(watcher, a, stopCh)
+		}(watcher, a, stopch)
 	}
 }
 
-func (a *AggregatedWatcher) closeWatcher(watcher Interface, stopCh *bcast.Member) {
-	watcher.Stop()
-	a.stopChGrp.Leave(stopCh)
+func (a *AggregatedWatcher) closeWatcher(watcher Interface) {
 	a.removeWatcherAndError(watcher)
 
-	if !a.allowWatcherReset && a.stopChGrp.MemberCount() == 0 {
-		//klog.Infof("Close watcher %v caused aggregated channel %v closed", watcher, a.aggChan)
-		a.stopChGrp.Close()
-		close(a.aggChan)
+	if !a.allowWatcherReset {
+		a.mapLock.RLock()
```

Comment:
if racing condition is a concern here, the Lock, Unlock() should be used instead of Block()